In [1]:
import pandas as pd
from neuronunit.optimization.optimization_management import public_opt
from sciunit.scores import RelativeDifferenceScore, ZScore
from sciunit.utils import config_set, config_get
config_set('PREVALIDATE', False)
from neuronunit.allenapi.allen_data_driven import make_allen_hard_coded_limited, make_allen_hard_coded_complete
import warnings
from neuronunit.optimization.model_parameters import BPO_PARAMS
warnings.filterwarnings("ignore")
_,_,_,a_cells = make_allen_hard_coded_limited()


[
    allen_suite_623960880,
    allen_suite_623893177,
    allen_suite471819401,
    allen_suite482493761,
    a_cells2,
    df
] = make_allen_hard_coded_complete()



/home/user/anaconda3/lib/python3.7/site-packages/cerberus/validator.py:1609: UserWarning: No validation schema is defined for the arguments of rule 'not_zero_obs_zscore'
  "'%s'" % method_name.split('_', 2)[-1]
test


# In this notebook
we fit Izhikevich and Adaptive Exponential Integrate and fire models to electrophysiology observations 
from the Allen Brain, using BluePyOpt and NeuronUnit.

# Immediately Below 
* is a list of some of the observations we will use:

In [2]:
df

TimeConstantTest       name RestingPotentialTest  \
623960880          23.8 ms  623960880             -65.1 mV   
623893177          27.8 ms  623893177             -77.0 mV   
471819401          13.8 ms  471819401             -77.5 mV   
482493761          24.4 ms  482493761             -71.6 mV   

          InputResistanceTest RheobaseTest  
623960880       241.0 megaohm      70.0 pA  
623893177       136.0 megaohm     190.0 pA  
471819401       132.0 megaohm     190.0 pA  
482493761       132.0 megaohm      70.0 pA

Allen Specimen ID: 471819401
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model AdExp

In [3]:
opt,obs_preds,df = public_opt(
    a_cells['471819401'],
    BPO_PARAMS,
    '471819401',
    "ADEXP",
    100,
    200,
    "IBEA",
    score_type=RelativeDifferenceScore
    )


gen	nevals	avg    	std    	min    	max 
1  	100   	1538.18	1514.29	12.4581	4000


2  	45    	1461.24	1480.68	9.93825	4000
3  	7     	400.301	484.527	9.93825	1009.87
4  	7     	236.369	414.534	9.93825	1010.11
5  	22    	340.159	467.049	8.89147	1011.75
6  	7     	250.161	424.425	8.89147	1008.4 
7  	13    	368.577	637.407	8.89147	4000   
8  	8     	69.951 	236.432	8.89147	1011.47
9  	11    	174.54 	368.894	8.89147	1009.28
10 	9     	338.45 	467.217	8.89147	1008.4 
11 	17    	204.008	393.457	8.69363	1008.97
12 	11    	381.581	481.446	8.69363	1007.46
13 	12    	188.616	382.059	8.27482	1008.41
14 	16    	292.199	449.188	8.11329	1006.76
15 	13    	352.12 	473.113	6.16534	1009.45
16 	12    	455.278	495.797	6.16534	1008.18
17 	12    	425.732	491.39 	6.16534	1007.7 
18 	16    	410.962	646.036	6.16534	4000   
19 	16    	187.02 	382.443	6.16534	1008.84
20 	6     	158.423	354.712	6.16534	1009.42
21 	14    	395.261	485.191	6.16534	1008.71
22 	17    	247.248	425.028	6.16534	1013.74
23 	15    	440.195	493.58 	6.16534	1007.42
24 	2     	439.454	493.7  	6.16534	1007.44
25 	18    	513

In [4]:
df

observations     predictions
RheobaseTest               190.0 pA       137.16 pA
TimeConstantTest            13.8 ms        14.22 ms
RestingPotentialTest       -77.5 mV       -77.42 mV
InputResistanceTest   132.0 megaohm  130.58 megaohm

# Solution Model Parameters:


In [5]:
opt.attrs

{'cm': 141.89465213418762,
 'v_spike': -56.87835077283676,
 'v_reset': -73.40688669611093,
 'v_rest': -77.61697031896016,
 'tau_m': 25.212126495261817,
 'a': 4.119094306368419,
 'b': 4.1830671821939305,
 'delta_T': 8.10331025873728,
 'tau_w': 231.5671836817075,
 'v_thresh': -46.06592482811786,
 'spike_delta': 52.60612763204166}

Allen Specimen ID: 471819401
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model Izhikevich

In [14]:
opt,obs_preds,df = public_opt(
    a_cells['471819401'],
    BPO_PARAMS,
    'fi_curve',
    "IZHI",
    100,
    360,
    "IBEA",
    )

gen	nevals	avg   	std    	min    	max 
1  	100   	1408.1	1110.01	13.2802	4000


2  	53    	1351.62	1096.69	12.0124	4000
3  	16    	741.547	609.985	12.0124	4000
4  	16    	355.01 	633.407	12.0124	4000
5  	16    	101.833	284.637	11.0153	1009.66
6  	18    	265.502	433.831	11.0153	1010.23
7  	13    	533.918	776.673	11.0153	4000   
8  	14    	116.998	304.791	11.0153	1010.18
9  	15    	325.726	625.535	11.0153	4000   
10 	14    	191.261	569.272	11.0153	4000   
11 	14    	55.9722	206.45 	10.2464	1009.92
12 	14    	220.312	405.507	10.2464	1010.13
13 	14    	667.255	633.794	10.2464	4000   
14 	11    	265.436	433.521	10.2464	1010.15
15 	19    	488.609	497.485	10.2464	1009.71
16 	14    	815.613	393.65 	10.2464	1009.84
17 	19    	756.232	604.448	10.2464	4000   
18 	13    	369.416	637.107	10.2464	4000   
19 	13    	56.3414	206.424	10.2464	1009.88
20 	11    	132.252	322.89 	10.2464	1009.57
21 	10    	280.174	441.937	10.2464	1010   
22 	21    	607.931	488.427	10.2464	1009.77
23 	9     	384.561	640.506	10.2464	4000   
24 	10    	86.4312	261.966	9.89462	1010.03
25 	19    	161.384	3

193	15    	663.299	471.708	8.18516	1007.14
194	12    	470.334	495.796	8.18516	1007.65
195	12    	381.386	481.16 	8.18516	1009.39
196	21    	663.226	472.553	7.6042 	1008.28
197	16    	411.441	488.028	7.6042 	1010.22
198	12    	618.83 	485.105	7.6042 	1010.1 
199	12    	678.831	629.604	7.6042 	4000   
200	10    	247.815	423.923	7.6042 	1009.01
201	18    	602.81 	487.956	7.6042 	1006.93
202	17    	484.75 	496.358	7.6042 	1007.75
203	15    	647.387	475.739	7.6042 	1009.9 
204	5     	261.837	432.054	7.6042 	1005.57
205	15    	513.9  	496.458	7.6042 	1007.71
206	10    	603.487	487.358	7.6042 	1006.45
207	8     	439.903	492.32 	7.6042 	1011.28
208	19    	455.267	493.364	7.6042 	1009.33
209	10    	292.663	447.795	7.6042 	1008.93
210	11    	366.371	476.426	7.6042 	1006.95
211	15    	396.224	484.465	7.6042 	1011.04
212	16    	544.093	495.468	7.6042 	1006.94
213	10    	543.403	495.265	7.6042 	1008.52
214	14    	321.378	461.028	7.6042 	1008.12
215	5     	603.246	486.805	7.6042 	1007.4 
216	15    	

In [8]:
df

observations    predictions
RheobaseTest               190.0 pA      349.62 pA
TimeConstantTest            13.8 ms       13.91 ms
RestingPotentialTest       -77.5 mV      -71.45 mV
InputResistanceTest   132.0 megaohm  129.1 megaohm

# Solution Parameters:

In [9]:
opt.attrs

{'C': 164.64073878392082,
 'k': 0.7824831664283994,
 'vr': -60.617527638867145,
 'vt': -44.83912380228744,
 'vPeak': 43.73174263796198,
 'a': 0.08560954808513295,
 'b': 13.193899560828427,
 'c': -56.381292959753345,
 'd': 131.64874325051048,
 'celltype': 3.3911289664348847}

# Just FI curve on AdExp model
Allen Specimin Id: 471819401

In [10]:
opt,obs_preds,df = public_opt(
    a_cells['fi_curve'],
    BPO_PARAMS,
    'fi_curve',
    "ADEXP",
    100,
    100,
    "IBEA"
    )

gen	nevals	avg    	std    	min    	max 
1  	100   	1651.54	495.397	4.32057	2000


2  	45    	1618.58	523.34 	4.32057	2000
3  	7     	1107.45	600.136	4.32057	2000
4  	7     	644.333	588.77 	4.32057	2000
5  	22    	377.121	592.899	4.32057	2000
6  	10    	49.0836	206.033	4.32057	1000.98
7  	18    	183.011	419.242	2.68177	2000   
8  	16    	19.0807	120.867	2.68177	1001.01
9  	11    	33.8411	169.501	2.68177	1000.19
10 	9     	152.272	395.12 	2.68177	2000   
11 	6     	17.6654	120.928	2.68177	1000.09
12 	20    	2.68177	0      	2.68177	2.68177
13 	16    	2.68177	0      	2.68177	2.68177
14 	13    	17.5697	120.95 	2.68177	1000.18
15 	11    	2.71294	0.253223	2.68177	4.77013
16 	17    	47.426 	269.104 	2.68177	2000   
17 	12    	17.5684	120.94  	2.68177	1000.09
18 	7     	47.4151	269.097 	2.68177	2000   
19 	20    	2.68177	0       	2.68177	2.68177
20 	16    	47.3895	269.094 	2.68177	2000   
21 	14    	92.1139	413.069 	2.68177	2000   
22 	13    	47.3791	269.057 	2.68177	2000   
23 	11    	32.4925	242.183 	2.68177	2000   
24 	19    	32.4925	242.183 	2.68177	2000   
25 	16    	2.

In [ ]:
df


In [ ]:
opt.attrs

Allen Specimen ID: 623960880
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model Izhikevich

In [13]:

opt,obs_preds,df = public_opt(
    allen_suite_623960880.tests,
    BPO_PARAMS,
    allen_suite_623960880.name,
    "IZHI",
    100,
    100,
    "IBEA"
    )

gen	nevals	avg    	std    	min    	max 
1  	100   	1408.55	1109.79	13.5395	4000


2  	53    	1352.04	1096.48	13.4028	4000
3  	16    	697.591	460.871	13.4028	1010.7
4  	16    	741.16 	440.257	13.4028	1010.14
5  	16    	533.576	495.996	13.4028	1010.01
6  	16    	651.509	476.398	13.4028	1010.85
7  	15    	666.81 	471.208	13.4028	1009.47
8  	16    	843.923	367.78 	13.4028	1009.61
9  	16    	784.843	413.426	13.1822	1010.73
10 	15    	681.169	466.115	13.1822	1009.47
11 	18    	903.729	303.824	13.1822	1010.18
12 	13    	859.146	354.055	13.1822	1009.46
13 	9     	858.562	354.091	13.1822	1009.15
14 	12    	992.221	120.514	13.1822	1009.09
15 	17    	932.83 	261.165	13.1822	1008.73
16 	6     	784.694	591.293	13.1822	4000   
17 	6     	294.844	447.684	13.1822	1008.59
18 	12    	739.553	440.249	13.1822	1009.97
19 	17    	976.451	168.972	13.1822	1008.03
20 	6     	917.221	283.53 	13.1822	1008.92
21 	15    	799.101	584.025	13.1822	4000   
22 	12    	383.564	480.07 	13.1822	1009.55
23 	20    	872.472	338.492	13.1822	1009.67
24 	12    	798.512	403.626	13.1822	1009.36
25 	21    	946.

In [15]:
df

observations     predictions
RheobaseTest               190.0 pA       128.27 pA
TimeConstantTest            13.8 ms        13.79 ms
RestingPotentialTest       -77.5 mV       -74.84 mV
InputResistanceTest   132.0 megaohm  131.95 megaohm

In [16]:
opt.attrs

{'C': 152.54814965173415,
 'k': 0.2500926432492604,
 'vr': -74.52543722085882,
 'vt': -48.78796029592446,
 'vPeak': 2.3478946326734205,
 'a': 0.026220704826477904,
 'b': 1.090819844829432,
 'c': -53.316065866270606,
 'd': 70.10644093197607,
 'celltype': 4.104128644144054}